src/dmd/dmodule.d

```diff
@@ -1355,3 +1382,221 @@ struct ModuleDeclaration
+that they've explicitly declared, however, if they provide no patterns or only provide
+"exclude patterns", then it is likely they want to include everything except modules
+that match one of their exclusions.
+
+Note that a user can always override this default behavior using the special '.' pattern.
+*/
+private __gshared bool includeByDefault = true;
+private __gshared Array!MatcherNode matchNodes;
+
+version(none)
+private void printMatcher(size_t matcherIndex)
+{
+    auto info = matchNodes[matcherIndex++];
+    if (info.isExclude)
+        printf("-");
+
+    if (info.depth == 0)
+        printf(".");
+    else
+    {
+        foreach (i; 0..info.depth)
+        {
+            if (i > 0) printf(".");
+            printf("%s", matchNodes[matcherIndex + i].id.toChars);
+        }
+    }
+}
+
+private void createMatchNodes()
+{
+    static size_t findSortedIndexToAddForDepth(size_t depth)
+    {
+        size_t index = 0;
+        for (; index < matchNodes.dim;)
+        {
+            auto info = matchNodes[index];
+            if (depth > matchNodes[index].depth)
+            {
+                break;
+            }
+            index += 1 + info.depth;
+        }
+        return index;
+    }
+
+    if (matchNodes.dim == 0)
+    {
+        for (size_t packageIndex = 0; packageIndex < global.params.includeMatchStrings.dim; packageIndex++)
+        {
+            auto matchString = global.params.includeMatchStrings[packageIndex];
+            auto depth = parseMatchStringDepth(matchString);
+            auto entryIndex = findSortedIndexToAddForDepth(depth);
+            matchNodes.split(entryIndex, depth + 1);
+            parseMatchString(matchString, &matchNodes[entryIndex], depth);
+            // if at least 1 "include pattern" is given, then it is assumed the
+            // user only wants to include modules that were explicitly given, which
+            // changes the default behavior from inclusion to exclusion.
+            if (includeByDefault && !matchNodes[entryIndex].isExclude)
+            {
+                //fprintf(global.stdmsg, "Matcher: found 'include pattern', switching default behavior to exclusion\n");
+                includeByDefault = false;
+            }
+        }
+
+        // Add the default 1 depth matchers
+        MatcherNode[8] defaultDepth1MatchNodes = [
+            MatcherNode(true, 1), MatcherNode(Id.std),
+            MatcherNode(true, 1), MatcherNode(Id.core),
+            MatcherNode(true, 1), MatcherNode(Id.etc),
+            MatcherNode(true, 1), MatcherNode(Id.object),
+        ];
+        {
+            auto index = findSortedIndexToAddForDepth(1);
+            matchNodes.split(index, defaultDepth1MatchNodes.length);
+            matchNodes.data[index .. index + defaultDepth1MatchNodes.length] = defaultDepth1MatchNodes[];
+        }
+    }
+}
+private bool endOfPattern(char c) { return c == ',' || c == '\0'; }
```

Comment:
these functions stuck together are surprising - why no empty line and hopefully a comment between them?